File: tools/exiftool-tables/capture_sony_main_conditions.pl
#!/usr/bin/env perl
# Evaluate `Sony::Main` model Conditions with a release's own ExifTool, in-process.
#
# `gen_sony_main_extra_tables.py` translates each `$$self{Model} =~ /RE/`
# Condition into an `MCond::ModelRe` the Rust interpreter evaluates with the
# `regex` crate. A Condition decides WHICH cameras get a tag, so a translation
# is only credible if it selects exactly the models the release's own Perl
# selects. This script is the Perl half of that proof; it never reads a
# version label to decide anything.
#
#   capture_sony_main_conditions.pl models <lib>
#       Print (JSON) the model strings named by <lib>'s in-memory
#       `%Image::ExifTool::Sony::Main` 0xb001 SonyModelID PrintConv.
#
#   capture_sony_main_conditions.pl eval <lib> <models.json> <tag-id>...
#       For every variant of every listed tag id whose Condition is exactly one
#       `$$self{Model} =~ /RE/` or `!~` test, `eval` that Condition string the
#       way ExifTool does (with `$self` a hash holding Model) over every model,
#       and print (JSON) the models it selects. Conditions of any other shape
#       are listed as `skipped` with their text, never evaluated; an id the
#       release does not define is listed in `absent_ids`.
use strict;
use warnings;
use JSON::PP;
use Digest::SHA qw(sha256_hex);

my ($mode, $lib, @rest) = @ARGV;
die "usage: $0 models <lib> | eval <lib> <models.json> <tag-id>...\n"
    unless defined $lib and ($mode // '') =~ /^(models|eval)$/;
die "$lib/Image/ExifTool/Sony.pm: not found\n" unless -f "$lib/Image/ExifTool/Sony.pm";
unshift @INC, $lib;
require Image::ExifTool;
require Image::ExifTool::Sony;
for my $mod ('Image/ExifTool.pm', 'Image/ExifTool/Sony.pm') {
    die "$mod loaded from $INC{$mod}, not $lib\n" unless $INC{$mod} eq "$lib/$mod";
}
my $main = do { no warnings "once"; \%Image::ExifTool::Sony::Main };
my $json = JSON::PP->new->canonical->pretty;

if ($mode eq 'models') {
    my $pc = $$main{0xb001}{PrintConv};
    die "Sony::Main 0xb001 has no PrintConv hash\n" unless ref $pc eq 'HASH';
    my %seen;
    for my $name (values %$pc) {
        # `ILCE-3000 / ILCE-3500`, `DSLR-A380/A390`, `DSLR-A850 (APS-C mode)`:
        # one PrintConv value can name several bodies; each is a Model string.
        (my $base = $name) =~ s/\s*\(.*\)\s*$//;
        my @parts = split m{\s*/\s*}, $base;
        my ($prefix) = $parts[0] =~ /^([A-Z]+-)/;
        for my $p (@parts) {
            $p = ($prefix // '') . $p unless $p =~ /-/;
            $seen{$p} = 1;
        }
    }
    print $json->encode({
        source => 'Sony::Main 0xb001 SonyModelID PrintConv',
        sony_pm_sha256 => sha256_file("$lib/Image/ExifTool/Sony.pm"),
        exiftool_version => "$Image::ExifTool::VERSION",
        models => [ sort keys %seen ],
    });
    exit 0;
}

my ($models_path, @ids) = @rest;
die "no tag ids\n" unless @ids;
my $models = decode_json(do { local (@ARGV, $/) = $models_path; <> });
die "$models_path: expected a JSON list of strings\n" unless ref $models eq 'ARRAY';

my (@conditions, @skipped, @absent);
for my $id_text (@ids) {
    my $id = $id_text =~ /^0x/i ? hex $id_text : $id_text;
    my $entry = $$main{$id};
    unless (defined $entry) {
        # Not this proof's concern: the generator refuses a missing id itself.
        push @absent, sprintf('0x%x', $id);
        next;
    }
    my @variants = ref $entry eq 'ARRAY' ? @$entry : ($entry);
    for my $i (0 .. $#variants) {
        my $v = $variants[$i];
        next unless ref $v eq 'HASH' and defined $$v{Condition};
        my $cond = $$v{Condition};
        my %row = (tag => sprintf('0x%x', $id), variant => $i, name => $$v{Name}, condition => $cond);
        unless ($cond =~ m{^\s*\$\$self\{Model\}\s*[=!]~\s*/[^/]*/\s*$}) {
            push @skipped, \%row;
            next;
        }
        my @selected;
        for my $model (@$models) {
            my $self = { Model => $model };
            my $hit = eval $cond;
            die "$row{tag} $$v{Name}: Condition died: $@" if $@;
            push @selected, $model if $hit;
        }
        $row{selects} = \@selected;
        push @conditions, \%row;
    }
}
print $json->encode({
    exiftool_version => "$Image::ExifTool::VERSION",
    sony_pm_sha256 => sha256_file("$lib/Image/ExifTool/Sony.pm"),
    perl => sprintf('%vd', $^V),
    conditions => \@conditions,
    skipped => \@skipped,
    absent_ids => \@absent,
});

sub sha256_file {
    my ($path) = @_;
    open my $fh, '<:raw', $path or die "$path: $!\n";
    local $/;
    return sha256_hex(<$fh>);
}

File: tools/exiftool-tables/capture_sony_main_conditions.py
#!/usr/bin/env python3
"""Record which Sony models each release's `Sony::Main` model Conditions select.

Writes `fixtures/sony_main_model_conditions.json`: for every release lib given,
every `$$self{Model} =~ /RE/` (or `!~`) Condition on a MANIFEST tag of
`gen_sony_main_extra_tables.py`, the models that release's own Perl selects
(evaluated by `capture_sony_main_conditions.pl`) next to the `MCond` text the
generator emits for that exact Condition. The Rust test
`main_extra::tests::model_conditions_select_what_each_releases_perl_selects`
evaluates each emitted `MCond` through the real interpreter over the same
models and requires the same selection; the Python tests require the emitted
text to still be what the generator produces.

The model list is the pinned release's 0xb001 SonyModelID PrintConv (real
Model strings), plus a few labelled synthetic probes aimed at the regexes'
edges (anchoring, case, a prefix of an alternative). Each release record's
`exiftool_version` is the loaded module's own `$Image::ExifTool::VERSION`,
recorded for the reader; nothing here admits or refuses on it.

usage:
  capture_sony_main_conditions.py --perl PERL --models-lib LIB13.59 \\
      --lib LIB [--lib LIB ...] -o fixtures/sony_main_model_conditions.json
"""
from __future__ import annotations

import argparse
import json
import subprocess
import sys
import tempfile
from pathlib import Path

import gen_sony_main_extra_tables as generator

HERE = Path(__file__).resolve().parent
PL = HERE / "capture_sony_main_conditions.pl"

# Not real bodies: edges the regexes must agree on anyway.
SYNTHETIC = [
    "",                 # empty Model
    "ilce-7m3",         # case: every pattern here is case-sensitive
    "XILCE-7M3",        # anchoring: `^` must hold
    "DSC-RX1RM3X",      # word-boundary edge for `\b` patterns
    "DSC-RX100M5AX",    # an alternative followed by more text (no `$`)
    "DSC-HX9",          # a strict prefix of an alternative
    "HX95",             # an inner alternative without its `DSC-` group
    "ZV-",              # a bare prefix
]


def perl(perl_bin, *args):
    proc = subprocess.run([perl_bin, str(PL), *args], capture_output=True, text=True)
    if proc.returncode != 0 or proc.stderr.strip():
        raise SystemExit(f"{PL.name} {args[0]} failed (rc={proc.returncode}): {proc.stderr.strip()}")
    return json.loads(proc.stdout)


def main():
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--perl", required=True)
    ap.add_argument("--models-lib", required=True)
    ap.add_argument("--lib", action="append", required=True)
    ap.add_argument("-o", "--output", required=True)
    args = ap.parse_args()

    model_src = perl(args.perl, "models", args.models_lib)
    models = model_src["models"] + [m for m in SYNTHETIC if m not in model_src["models"]]
    ids = [f"0x{i:x}" for i in generator.MANIFEST]

    releases = []
    with tempfile.NamedTemporaryFile("w", suffix=".json") as tmp:
        json.dump(models, tmp)
        tmp.flush()
        for lib in args.lib:
            rel = perl(args.perl, "eval", lib, tmp.name, *ids)
            for row in rel["conditions"]:
                row["generated"] = generator.translate_cond(int(row["tag"], 16), row["name"], row["condition"])
            releases.append(rel)

    out = {
        "about": "Models each release's own Perl selects with each Sony::Main model Condition; "
                 "regenerate with capture_sony_main_conditions.py.",
        "models": {"source": f"{model_src['source']} ({model_src['exiftool_version']}, "
                             f"Sony.pm sha256 {model_src['sony_pm_sha256']})",
                   "real": model_src["models"], "synthetic": SYNTHETIC},
        "releases": releases,
    }
    Path(args.output).write_text(json.dumps(out, indent=1, sort_keys=True) + "\n", encoding="utf-8")


if __name__ == "__main__":
    try:
        main()
    except generator.Unsupported as e:
        print(f"capture_sony_main_conditions.py: {e}", file=sys.stderr)
        sys.exit(1)
